### PR TITLE
Allow destructuring in ambient

### PR DESCRIPTION
**Pull request by yuit**
Address #2421

### CODE REVIEW

**Comment by DanielRosenwasser** [comment]
There are no tests here for default initializers.

**Comment by DanielRosenwasser** [comment]
Consider a few tests like:

```TypeScript
function f({x = 10, y: [a,b,c,d] = [1,2,3,4]} = { x: 10, y: [2,4,6,8]) {
}
```

**Comment by JsonFreeman** [comment]
Yes, make sure that default initializers are disallowed. Also make sure that rest elements are allowed.

**Comment by JsonFreeman** [comment]
:+1:

**Comment by JsonFreeman** [comment]
:tada:

File: src/compiler/declarationEmitter.ts [hunk 1]
```diff
@@ -1353,30 +1361,84 @@ module ts {
...
+
+                    // If bindingElement has propertyName property, then its name must be another bindingPattern of SyntaxKind.ObjectBindingPattern
+                    emitBindingPattern(<BindingPattern>bindingElement.name);
+                }
+                else if (bindingElement.name) {
+                    if (isBindingPattern(bindingElement.name)) {
+                        // If it is a nested binding pattern, we will recursively descend into each element and emit each one separately.
+                        // In the case of rest element, we will omit rest element.
+                        // Example:
+                        //      original: function foo([a, [[b]], c] = [1,[["string"]], 3]) {}
```
Comment (CyrusNajmabadi) [inline src/compiler/declarationEmitter.ts]: Great comments, thanks!

**Pull request merged**